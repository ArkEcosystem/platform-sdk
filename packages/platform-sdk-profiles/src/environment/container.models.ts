export const Identifiers = {
	AppData: "Data<App>",
	Coins: "Coins",
	ContactRepository: "ContactRepository",
	DataRepository: "DataRepository",
	HttpClient: "HttpClient",
	Migrator: "Migrator",
	ProfileRepository: "ProfileRepository",
	SettingRepository: "SettingRepository",
	Storage: "Storage",
	WalletRepository: "WalletRepository",
};
