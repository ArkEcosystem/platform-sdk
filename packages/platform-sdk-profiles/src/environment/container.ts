export class Container {
	readonly #bindings: Map<string, any> = new Map();

	public get<T>(key: string): T {
		return this.#bindings.get(key);
	}

	public set(key: string, value: unknown): void {
		this.#bindings.set(key, value);
	}

	public has(key: string): boolean {
		return this.#bindings.has(key);
	}

	public forget(key: string): void {
		this.#bindings.delete(key);
	}
}

export const container = new Container();
