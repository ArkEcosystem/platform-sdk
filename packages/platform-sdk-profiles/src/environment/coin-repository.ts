import { Coins, Contracts } from "@arkecosystem/platform-sdk";

import { DataRepository } from "../repositories/data-repository";
import { makeCoin } from "./container.helpers";

export class CoinRepository {
	readonly #dataRepository: DataRepository = new DataRepository();

	public delegates(coin: string, network: string): any {
		const result = this.#dataRepository.get(`${coin}.${network}.delegates`);

		if (result === undefined) {
			throw new Error(
				`The delegates for [${coin}.${network}] have not been synchronized yet. Please call [syncDelegates] before using this method.`,
			);
		}

		return result;
	}

	public fees(coin: string, network: string, days = 7): any {
		const result = this.#dataRepository.get(`${coin}.${network}.fees.${days}`);

		if (result === undefined) {
			throw new Error(
				`The delegates for [${coin}.${network}.fees.${days}] have not been synchronized yet. Please call [syncFees] before using this method.`,
			);
		}

		return result;
	}

	public async syncDelegates(coin: string, network: string): Promise<void> {
		const instance: Coins.Coin = await makeCoin(coin, network);
		const instanceKey = `${coin}.${network}.delegates`;

		let result: Contracts.WalletData[] = [];
		let hasMore = true;
		// TODO: use the nextPage() method as cursor like aggregates
		let cursor = 1;

		while (hasMore) {
			const response = await instance.client().delegates({ cursor });

			result = result.concat(response.items());

			hasMore = response.hasMorePages();

			cursor++;
		}

		this.#dataRepository.set(
			instanceKey,
			result.map((delegate: Contracts.WalletData) => delegate.toObject()),
		);
	}

	public async syncFees(coin: string, network: string, days = 7): Promise<void> {
		const instance: Coins.Coin = await makeCoin(coin, network);

		this.#dataRepository.set(`${coin}.${network}.fees.${days}`, await instance.fee().all(days));
	}
}
