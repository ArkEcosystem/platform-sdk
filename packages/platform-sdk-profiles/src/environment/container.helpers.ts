import { Coins } from "@arkecosystem/platform-sdk";

import { container } from "./container";
import { Identifiers } from "./container.models";

export const makeCoin = async (coin: string, network: string): Promise<Coins.Coin> =>
	Coins.CoinFactory.make(container.get<Coins.CoinSpec>(Identifiers.Coins)[coin.toUpperCase()], {
		network,
		httpClient: container.get(Identifiers.HttpClient),
	});
