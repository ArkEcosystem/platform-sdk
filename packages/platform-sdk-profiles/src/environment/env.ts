import { Coins } from "@arkecosystem/platform-sdk";
import Joi from "joi";

import {
	IDataRepository,
	IDelegateService,
	IExchangeRateService,
	IFeeService,
	IKnownWalletService,
	IPluginRegistry,
	IProfileRepository,
	IWalletService,
} from "../contracts";
import { DriverFactory } from "../drivers/driver.factory";
import { container } from "./container";
import { Identifiers } from "./container.models";
import { CoinList, EnvironmentOptions, Storage, StorageData } from "./env.models";

export class Environment {
	private storage: StorageData | undefined;

	public constructor(options: EnvironmentOptions) {
		this.configureDriver(options);
	}

	/**
	 * Verify the integrity of the storage.
	 *
	 * @param {StorageData} { data, profiles }
	 * @returns {Promise<void>}
	 * @memberof Environment
	 */
	public async verify(storage?: StorageData): Promise<void> {
		if (storage === undefined) {
			storage = await container.get<Storage>(Identifiers.Storage).all<StorageData>();
		}

		const data: object = storage.data || {};
		const profiles: object = storage.profiles || {};

		const { error, value } = Joi.object({
			data: Joi.object().required(),
			profiles: Joi.object().pattern(Joi.string().uuid(), Joi.object()).required(),
		}).validate({ data, profiles });

		if (error) {
			throw new Error(`Terminating due to corrupted state: ${error}`);
		}

		this.storage = value;
	}

	/**
	 * Load the data from the storage.
	 *
	 * This has to be manually called and should always be called before booting
	 * of the environment instance. This will generally be only called on application start.
	 *
	 * @returns {Promise<void>}
	 * @memberof Environment
	 */
	public async boot(): Promise<void> {
		if (this.storage === undefined) {
			throw new Error("Please call [verify] before booting the environment.");
		}

		if (Object.keys(this.storage.data).length > 0) {
			this.data().fill(this.storage.data);
		}

		if (Object.keys(this.storage.profiles).length > 0) {
			this.profiles().fill(this.storage.profiles);
		}

		/* istanbul ignore next */
		if (container.has(Identifiers.ExchangeRateService)) {
			await container.get<IExchangeRateService>(Identifiers.ExchangeRateService).restore();
		}
	}

	/**
	 * Save the data to the storage.
	 *
	 * This has to be manually called and should always be called before disposing
	 * of the environment instance. For example on application shutdown or when switching profiles.
	 *
	 * @returns {Promise<void>}
	 * @memberof Environment
	 */
	public async persist(): Promise<void> {
		const storage: Storage = container.get<Storage>(Identifiers.Storage);

		await storage.set("profiles", this.profiles().toObject());

		await storage.set("data", this.data().all());
	}

	/**
	 * Access the application data.
	 *
	 * @returns {DataRepository}
	 * @memberof Environment
	 */
	public data(): IDataRepository {
		return container.get(Identifiers.AppData);
	}

	/**
	 *
	 *
	 * @returns {DelegateService}
	 * @memberof Environment
	 */
	public delegates(): IDelegateService {
		return container.get(Identifiers.DelegateService);
	}

	/**
	 * Access the exchange rate service.
	 *
	 * @returns {ExchangeRateService}
	 * @memberof Environment
	 */
	public exchangeRates(): IExchangeRateService {
		return container.get(Identifiers.ExchangeRateService);
	}

	/**
	 * Access the fees service.
	 *
	 * @returns {FeeService}
	 * @memberof Environment
	 */
	public fees(): IFeeService {
		return container.get(Identifiers.FeeService);
	}

	/**
	 * Access the known wallets service.
	 *
	 * @returns {KnownWalletService}
	 * @memberof Environment
	 */
	public knownWallets(): IKnownWalletService {
		return container.get(Identifiers.KnownWalletService);
	}

	/**
	 * Access the plugin registry.
	 *
	 * @returns {IPluginRegistry}
	 * @memberof Environment
	 */
	public plugins(): IPluginRegistry {
		return container.get(Identifiers.PluginRegistry);
	}

	/**
	 * Access the profile repository.
	 *
	 * @returns {ProfileRepository}
	 * @memberof Environment
	 */
	public profiles(): IProfileRepository {
		return container.get(Identifiers.ProfileRepository);
	}

	/**
	 * Access the wallet service.
	 *
	 * @returns {WalletService}
	 * @memberof Environment
	 */
	public wallets(): IWalletService {
		return container.get(Identifiers.WalletService);
	}

	/**
	 * Register a new coin implementation by its ticker, for example ARK or BTC.
	 *
	 * @param {string} coin
	 * @param {Coins.CoinSpec} spec
	 * @memberof Environment
	 */
	public registerCoin(coin: string, spec: Coins.CoinSpec): void {
		if (container.get<CoinList>(Identifiers.Coins)[coin]) {
			throw new Error(`The coin [${coin}] is already registered.`);
		}

		container.get<CoinList>(Identifiers.Coins)[coin] = spec;
	}

	/**
	 * Return a list of all available networks.
	 *
	 * @returns {Coins.Network[]}
	 * @memberof Environment
	 */
	public availableNetworks(): Coins.Network[] {
		const coins: CoinList = container.get<CoinList>(Identifiers.Coins);

		const result: Coins.Network[] = [];

		for (const coin of Object.values(coins)) {
			for (const network of Object.values(coin.manifest.networks)) {
				result.push(new Coins.Network(coin.manifest, network));
			}
		}

		return result;
	}

	/**
	 * Remove all bindings from the container and optionally rebind them.
	 *
	 * @memberof Environment
	 */
	public reset(options?: EnvironmentOptions): void {
		container.flush();

		if (options !== undefined) {
			this.configureDriver(options);
		}
	}

	/**
	 * Set the migrations that should be used for profiles, if applicable.
	 *
	 * @param {object} schemas
	 * @param {string} version
	 * @memberof Environment
	 */
	public setMigrations(schemas: object, version: string): void {
		container.bind(Identifiers.MigrationSchemas, schemas);
		container.bind(Identifiers.MigrationVersion, version);
	}

	/**
	 * Create a driver instance and all necessary container bindings.
	 *
	 * @memberof Environment
	 */
	private configureDriver(options: EnvironmentOptions): void {
		return DriverFactory.make("memory", container, options);
	}
}
