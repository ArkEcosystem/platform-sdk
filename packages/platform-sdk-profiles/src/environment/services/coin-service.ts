import { Coins } from "@arkecosystem/platform-sdk";

import { DataRepository } from "../../repositories/data-repository";
import { container } from "../container";
import { Identifiers } from "../container.models";

export class CoinService {
	readonly #dataRepository: DataRepository = new DataRepository();

	public all(): Record<string, Coins.Coin> {
		return this.#dataRepository.all() as Record<string, Coins.Coin>;
	}

	public values(): Coins.Coin[] {
		return this.#dataRepository.values();
	}

	public entries(): [string, string[]][] {
		const result: Record<string, string[]> = {};

		for (const [coin, networks] of Object.entries(this.all())) {
			result[coin] = [];

			for (const [network, children] of Object.entries(networks)) {
				if (children !== undefined) {
					for (const child of Object.keys(children)) {
						result[coin].push(`${network}.${child}`);
					}
				} else {
					result[coin].push(network);
				}
			}
		}

		return Object.entries(result);
	}

	public get(coin: string, network: string): Coins.Coin {
		const instance: Coins.Coin | undefined = this.#dataRepository.get<Coins.Coin>(`${coin}.${network}`);

		if (instance === undefined) {
			throw new Error(`An instance for [${coin}.${network}] does not exist.`);
		}

		return instance;
	}

	public async push(coin: string, network: string, options: object = {}, useForce = false): Promise<Coins.Coin> {
		if (!useForce && this.has(coin, network)) {
			return this.get(coin, network);
		}

		this.#dataRepository.set(
			`${coin}.${network}`,
			await Coins.CoinFactory.make(container.get<Coins.CoinSpec>(Identifiers.Coins)[coin.toUpperCase()], {
				network,
				httpClient: container.get(Identifiers.HttpClient),
				...options,
			}),
		);

		return this.get(coin, network);
	}

	public has(coin: string, network: string): boolean {
		return this.#dataRepository.has(`${coin}.${network}`);
	}
}
