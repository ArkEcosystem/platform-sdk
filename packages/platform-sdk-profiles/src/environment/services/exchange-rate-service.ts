import { MarketService } from "@arkecosystem/platform-sdk-markets";

import { Profile } from "../../profiles/profile";
import { ProfileSetting } from "../../profiles/profile.models";
import { ProfileRepository } from "../../repositories/profile-repository";
import { ReadWriteWallet, WalletData } from "../../wallets/wallet.models";
import { container } from "../container";
import { Identifiers } from "../container.models";

export class ExchangeRateService {
	public async syncAll(): Promise<void> {
		const profiles: Profile[] = container.get<ProfileRepository>(Identifiers.ProfileRepository).values();

		const promises: Promise<void>[] = [];
		for (const profile of profiles) {
			for (const [currency, wallets] of Object.entries(profile.wallets().allByCoin())) {
				promises.push(this.syncCoinByProfile(profile, currency, Object.values(wallets)));
			}
		}

		await Promise.allSettled(promises);
	}

	public async syncCoinByProfile(profile: Profile, currency: string, wallets?: ReadWriteWallet[]): Promise<void> {
		if (wallets === undefined) {
			wallets = profile
				.wallets()
				.values()
				.filter((wallet: ReadWriteWallet) => wallet.currency() === currency && wallet.network().isLive());
		} else {
			wallets = wallets.filter((wallet: ReadWriteWallet) => wallet.network().isLive());
		}

		if (!wallets.length) {
			return;
		}

		const marketService = MarketService.make(
			profile.settings().get(ProfileSetting.MarketProvider) || "coingecko",
			container.get(Identifiers.HttpClient),
		);

		const exchangeCurrency: string = profile.settings().get(ProfileSetting.ExchangeCurrency) || "BTC";
		const exchangeRate = await marketService.dailyAverage(currency, exchangeCurrency, +Date.now());

		for (const wallet of wallets) {
			wallet.data().set(WalletData.ExchangeCurrency, exchangeCurrency);
			wallet.data().set(WalletData.ExchangeRate, exchangeRate);
		}
	}
}
