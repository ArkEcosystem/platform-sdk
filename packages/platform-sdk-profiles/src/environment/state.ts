import { Exceptions } from "@arkecosystem/platform-sdk";

import { IProfile } from "../contracts";
import { container } from "./container";
import { State as Identifiers } from "./container.models";

export class State {
	/**
	 * Get or set the active profile.
	 *
	 * @static
	 * @param {IProfile} [instance]
	 * @return {*}  {IProfile}
	 * @memberof State
	 */
	public static profile(instance?: IProfile): IProfile {
		if (instance !== undefined) {
			if (container.has(Identifiers.Profile)) {
				container.rebind(Identifiers.Profile, instance);
			} else {
				container.bind(Identifiers.Profile, instance);
			}
		}

		const profile: IProfile | undefined = container.get(Identifiers.Profile);

		if (profile === undefined) {
			throw new Exceptions.BadStateException("profile", "There is no active profile");
		}

		return profile;
	}

	/**
	 * Forget the given key.
	 *
	 * @static
	 * @param {string} key
	 * @memberof State
	 */
	public static forget(key: string): void {
		if (container.missing(Identifiers.Profile)) {
			throw new Exceptions.BadStateException("forget", "There is no active profile");
		}

		container.unbind(key);
	}
}
