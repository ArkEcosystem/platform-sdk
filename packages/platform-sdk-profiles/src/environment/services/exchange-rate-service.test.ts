import "jest-extended";

import { ARK } from "@arkecosystem/platform-sdk-ark";
import { Request } from "@arkecosystem/platform-sdk-http-got";
import nock from "nock";

import { identity } from "../../../test/fixtures/identity";
import { container } from "../../environment/container";
import { Identifiers } from "../../environment/container.models";
import { Profile } from "../../profiles/profile";
import { ProfileSetting } from "../../profiles/profile.models";
import { ProfileRepository } from "../../repositories/profile-repository";
import { ReadWriteWallet, WalletData } from "../../wallets/wallet.models";
import { CoinService } from "./coin-service";
import { ExchangeRateService } from "./exchange-rate-service";

let profile: Profile;
let wallet: ReadWriteWallet;
let subject: ExchangeRateService;

let liveSpy: jest.SpyInstance;
let testSpy: jest.SpyInstance;

beforeEach(async () => {
	nock.cleanAll();

	nock(/.+/)
		.get("/api/node/configuration")
		.reply(200, require("../../../test/fixtures/client/configuration.json"))
		.get("/api/peers")
		.reply(200, require("../../../test/fixtures/client/peers.json"))
		.get("/api/node/configuration/crypto")
		.reply(200, require("../../../test/fixtures/client/cryptoConfiguration.json"))
		.get("/api/node/syncing")
		.reply(200, require("../../../test/fixtures/client/syncing.json"))
		.get("/api/wallets/D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib")
		.reply(200, require("../../../test/fixtures/client/wallet.json"))
		.get("/api/delegates")
		.reply(200, require("../../../test/fixtures/client/delegates-1.json"))
		.get("/api/delegates?page=2")
		.reply(200, require("../../../test/fixtures/client/delegates-2.json"))
		.persist();

	const profileRepository = new ProfileRepository();

	container.set(Identifiers.HttpClient, new Request());
	container.set(Identifiers.CoinService, new CoinService());
	container.set(Identifiers.Coins, { ARK });
	container.set(Identifiers.ProfileRepository, profileRepository);

	profile = profileRepository.create("John Doe");
	wallet = await profile.wallets().importByMnemonic(identity.mnemonic, "ARK", "ark.devnet");

	liveSpy = jest.spyOn(wallet.network(), "isLive").mockReturnValue(true);
	testSpy = jest.spyOn(wallet.network(), "isTest").mockReturnValue(false);

	subject = new ExchangeRateService();
});

afterEach(() => {
	liveSpy.mockRestore();
	testSpy.mockRestore();
});

beforeAll(() => nock.disableNetConnect());

it("should sync the exchange rate for ARK to BTC", async () => {
	profile.settings().set(ProfileSetting.MarketProvider, "cryptocompare");

	nock(/.+/)
		.get("/data/dayAvg")
		.query(true)
		.reply(200, { BTC: 0.00005048, ConversionType: { type: "direct", conversionSymbol: "" } })
		.persist();

	await subject.syncAll();

	expect(wallet.data().get(WalletData.ExchangeRate)).toBe(0.00005048);
});

describe("#syncCoinByProfile", () => {
	it("should sync a coin for specific profile with wallets argument", async () => {
		profile.settings().set(ProfileSetting.MarketProvider, "cryptocompare");

		nock(/.+/)
			.get("/data/dayAvg")
			.query(true)
			.reply(200, { BTC: 0.00005048, ConversionType: { type: "direct", conversionSymbol: "" } })
			.persist();

		await subject.syncCoinByProfile(
			profile,
			"DARK",
			profile
				.wallets()
				.values()
				.filter((wallet: ReadWriteWallet) => wallet.currency() === "DARK"),
		);

		expect(wallet.data().get(WalletData.ExchangeRate)).toBe(0.00005048);
	});

	it("should sync a coin for specific profile without wallets argument", async () => {
		profile.settings().set(ProfileSetting.MarketProvider, "cryptocompare");

		nock(/.+/)
			.get("/data/dayAvg")
			.query(true)
			.reply(200, { BTC: 0.00002134, ConversionType: { type: "direct", conversionSymbol: "" } })
			.persist();

		await subject.syncCoinByProfile(profile, "DARK");

		expect(wallet.data().get(WalletData.ExchangeRate)).toBe(0.00002134);
	});
});
