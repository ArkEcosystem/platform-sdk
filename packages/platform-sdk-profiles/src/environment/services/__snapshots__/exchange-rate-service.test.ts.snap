// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeRateService should sync a coin for specific profile with wallets argument 1`] = `
Object {
  "EXCHANGE_RATE_SERVICE": Object {
    "DARK": Object {
      "BTC": Object {
        "2020-01-04": 0.00001978,
        "2020-01-05": 0.00001979,
        "2020-01-06": 0.00001963,
        "2020-01-07": 0.00001791,
        "2020-01-08": 0.00001776,
        "2020-01-09": 0.00001766,
        "2020-01-10": 0.00001756,
        "2020-01-11": 0.00001753,
        "2020-01-12": 0.00001711,
        "2020-01-13": 0.00001717,
        "2020-01-14": 0.0000161,
        "2020-01-15": 0.00001686,
        "2020-01-16": 0.00001665,
        "2020-01-17": 0.00001749,
        "2020-01-18": 0.00001796,
        "2020-01-19": 0.00001789,
        "2020-01-20": 0.00001764,
        "2020-01-21": 0.00001809,
        "2020-01-22": 0.00001899,
        "2020-01-23": 0.0000184,
        "2020-01-24": 0.00001939,
        "2020-01-25": 0.0000189,
        "2020-01-26": 0.00001816,
        "2020-01-27": 0.00001765,
        "2020-01-28": 0.00001733,
        "2020-01-29": 0.00001697,
        "2020-01-30": 0.00001684,
        "2020-01-31": 0.00001713,
        "2020-02-01": 0.00001779,
        "2020-02-02": 0.00001867,
        "2020-02-03": 0.00001904,
        "2020-12-15": 0.00005048,
      },
    },
  },
  "data": Object {},
  "profiles": Object {},
}
`;
